File: csharpcore/GildedRose/GildedRoseItem.cs
using GildedRoseKata.Interfaces;
using GildedRoseKata.ItemTypes;

namespace GildedRoseKata;

/**
 * Adapter Pattern, used to convert an object to another with better compatibility with the application being written or refactored.
 * Original object remains intact and the adapter hides all the logic of the conversion.
 * In this case, object persistence is used to update the original object within it's collection passed into the GildedRose Class.
 */

public class GildedRoseItem : IGildedRoseItem
{
  private Item Item { get; }

  public string Name => Item.Name;
  public int Quality { get => Item.Quality; set => Item.Quality = value; }
  public int SellIn { get => Item.SellIn; set => Item.SellIn = value; }
  public ItemType ItemType { get; set; } = null!;
  
  public GildedRoseItem(Item item)
  {
    Item = item;
    IdentifyItem();
  }

  /**
   * Item identification, making use of the State Pattern to pattern match the names of each object to their relevant types.
   * The type contains identification characteristics and unique logic constraints to be applied to the item.
   * Additional types can be added as new classes, inheriting the base characteristics from the ItemType Class. 
   */
  private void IdentifyItem()
  {
    switch (Name)
    {
      case string legendary when legendary.ToLower().Contains("sulfuras"):
        ItemType = new LegendaryItemType();
        break;
      case string cheese when cheese.ToLower().Contains("aged brie"):
        ItemType = new CheeseItemType();
        break;
      case string conjured when conjured.ToLower().Contains("conjured"):
        ItemType = new ConjuredItemType();
        break;
      case string pass when pass.ToLower().Contains("backstage passes"):
        ItemType = new ConcertPassItemType();
        break;
      default:
        ItemType = new GeneralItemType();
        break;
    }
  }

  public void Update()
  {
    int quality = Quality;
    int sellIn = SellIn;
    
    ItemType.UpdateQuality(ref quality, ref sellIn);

    Quality = quality;
    SellIn = sellIn;
  }
}

File: ruby/gilded_rose_item.rb
require File.join(File.dirname(__FILE__), 'ItemTypes/general_item_type')
require File.join(File.dirname(__FILE__), 'ItemTypes/legendary_item_type')
require File.join(File.dirname(__FILE__), 'ItemTypes/cheese_item_type')
require File.join(File.dirname(__FILE__), 'ItemTypes/conjured_item_type')
require File.join(File.dirname(__FILE__), 'ItemTypes/concert_pass_item_type')

class GildedRoseItem
  attr_accessor :item, :item_type

  def initialize(item)
    @item = item
    identify
  end

  def name
    @item.name
  end

  def quality
    @item.quality
  end

  def quality=(value)
    @item.quality = value
  end

  def sell_in
    @item.sell_in
  end

  def sell_in=(value)
    @item.sell_in = value
  end

  def identify
    @item_type = GeneralItemType.new
    @item_type = LegendaryItemType.new if name.downcase.include? "sulfuras"
    @item_type = CheeseItemType.new if name.downcase.include? "aged brie"
    @item_type = ConjuredItemType.new if name.downcase.include? "conjured"
    @item_type = ConcertPassItemType.new if name.downcase.include? "backstage passes"
  end

  def update
    qual = quality
    sell = sell_in

    qual, sell = @item_type.update_quality(qual,sell)

    quality = qual
    sell_in = sell
  end
end
